[6.2] [Performance] Access Rules smart preload, reduce memory and time use, for large sites

Fedik opened the pull request:
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/43649.

### Summary of Changes

Currently while checking access, the `Access::preload()` pulling everything from the access table.
It is neglectable for small sites, however for large sites it has huge impact on the site performance.

I added a new method for "smart preload", which allows to preload only needed items. And applied this method for most of the com_content views, where it make sense. 

#### It works as following:

In the View class (or other places before access check) developer call to pre-load access for loaded items with (example for `com_content`):
```php
$assetsList = ['com_content.article.1', 'com_content.article.2', 'com_content.category.1'];
Access::preloadItems('com_content',  $assetsList);
```
Where the `$assetsList` is a list for asset name for `$items`.

This will load access information for these items and its parent in to memory. 
And each following call of `$user->authorise('core.xxx', $assetName)` will reuse preloaded item.


### Testing Instructions

Create a site with 20k articles (more is better). Can use following plugin to generate random content https://github.com/Fedik/plg_sampledata_bigdata

Enable debug.
Open article detail or blog view.
Look for memory usage in the debug bar, and for time.


### Actual result BEFORE applying this Pull Request
In my test:
memory usage around 45Mb
time 200-300ms


### Expected result AFTER applying this Pull Request
In my test:
memory usage around ~12Mb 
time ~100ms


### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed
- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

Fedik committed:
Access::preloadItems

Fedik committed:
Access::preloadItems

Fedik committed:
Access::preloadItems

Fedik committed:
Access::preloadItems

Fedik committed:
Access::preloadItems lft, rgt

Fedik committed:
Access::preloadItems parents

Fedik committed:
Access::preloadItems parents2

Fedik committed:
Revert " Access::preloadItems parents2"

This reverts commit 02a10187913762391c04906d840431f93fa92245.

Fedik committed:
Access::preloadItems

Fedik committed:
Access::preloadItems

Fedik committed:
Access::preloadItems

Fedik committed:
Access::preloadItems

Fedik commented:
Will rebase it after upmarge

Fedik commented on libraries/src/Access/Access.php:
```diff
@@ -209,6 +209,60 @@ public static function preload($assetTypes = 'components', $reload = false)
+            ->where(
+                $db->quoteName($key) . ' IN ('
+                . implode(',', $query->bindArray($assetsList, $bindAs)) . ')'
+            );
```
@richard67 what do you think, it is possible to query parents with lft, rgt, but fast? :smiley: 
I mean I tried :
```sql
SELECT DISTINCT b.id, b.name, b.parent_id
FROM bd5_assets AS a
LEFT JOIN bd5_assets AS b ON b.lft <= a.lft AND b.rgt >= a.rgt
WHERE a.name IN ('com_content.article.24');
```
But it is much slower than doing 2-4 queries, in my test:
- lft,rgt: 200-300ms
- multiple queries: a few queries around 50-100μs

richard67 replied:
Why using lft and rgt? Can't you rely on the parent_id? What are you trying to do?

And with the current code, why using where and implode and not using the array directly with whereIn?

Fedik replied:
> What are you trying to do?

Preloading the assets by given name or id. Also this need to preload all their parents. 
Curently I am doing it in recursion, however if there a smarter way to make it fast in 1-2 query, would be cooler.

> Can't you rely on the parent_id?

I was unable to make it work in 1 query.

Fedik replied:
> why using where and implode and not using the array directly with whereIn?

I copied it from somewhere in Joomla. Did not seen how to use the array directly.
I will check again.

richard67 replied:
I see. All parents up to the root.

richard67 replied:
Regarding whereIn with array:
```
$query->whereIn($db->quoteName($key), $assetsList, $bindAs);
```

Fedik replied:
> Regarding whereIn with array

Updated, thanks.

richard67 replied:
Regarding the slow query for all in one: Maybe it is slow because we have only a combined index on lft and rgt but not a separate index for each of them.

Is it faster if you add these indexes?

```
ALTER TABLE `#__assets` ADD INDEX `idx_lft` (`lft`), ADD INDEX `idx_rgt` (`rgt`);
```

richard67 replied:
Using one query for all with the parent_id would work with a so-called recursive query, but as our database query doesn't have methods for that it would either need the same syntax for MySQL/MariaDB and PostgreSQL or different hard-coded queries for the different database type like we have it here or there e.g. in com_finder.

Fedik replied:
> Is it faster if you add these indexes?

I will test.

richard67 replied:
@Fedik Regarding recursive query: The following works on MySQL >= 8.0. I haven't checked MariaDB and PostgreSQL yet.
```
WITH RECURSIVE `assets` (`id`, `name`, `parent_id`, `level`, `rules`) AS (
  SELECT `id`, `name`, `parent_id`, `level`, `rules`
  FROM `bd5_assets `
  WHERE `name` IN ('com_languages.language.1', 'com_modules.module.86')
  UNION ALL
  SELECT a.`id`, a.`name`, a.`parent_id`, a.`level`, a.`rules`
  FROM `bd5_assets ` a
  INNER JOIN `assets` on a.`id` = `assets`.`parent_id`
) SELECT DISTINCT `id`, `name`, `parent_id`, `level`, `rules` FROM `assets` ORDER BY `level`;
```
But to me it seems that's not really faster.

Fedik replied:
Hm, you know, it is much faster than "left join with lft/rgt": ~250ms for JOIN vs ~845μs for RECURSIVE.

But still a bit slower than 1-2 Query for PHP recursive call, I mean, when there 1-2 lvl of nesting. For deeper nesting it probably will be faster.

Thank you @richard67 for checking!

I think will stop on current implementation with PHP recursive call. 
I have a hope no one have nesting more than 10 level deep :smile:

richard67 replied:
I am ok with that.

Hackwar replied:
Just for clarity: every request for an article already contains at least 3 levels. The component, the category and the article itself. A nesting of 5 categories depths isn't really that strange, resulting in 7 levels depth.

richard67 replied:
plus the root asset at the top.

richard67 replied:
According to https://mariadb.com/kb/en/with/ , recursive common table expressions are supported since MariaDB version 10.2.2, and the syntax seems to be the same as for MySQL >= 8.0, so my recursive query should work with MySQL and MariaDB for J5.

richard67 replied:
Will check later for PostgreSQL.

Fedik replied:
The component and root preloaded by  https://github.com/joomla/joomla-cms/blob/21f6a2b5d14bea553978b2cac88a0cead0754912/libraries/src/Access/Access.php#L526

It is safe to ignore them.

A new `preloadItems()` will preload only missing elements, it uses `array_diff()` to check what already in the memory cache

richard67 replied:
For PostgreSQL recursive common table expressions seem to be supported since version 14: https://www.postgresql.org/docs/14/queries-with.html#QUERIES-WITH-RECURSIVE . But for J5 we have 12 as minimum version.

Fedik replied:
More I think about it, more I like PHP recursion. Because it allows to check what already was preloaded.
Example we have: 
```
category.1 
- category.2 
-- article.1
-- article.2
-- article.3
-- article.4
```
On preload `['article.1', 'article.3']` will be preloaded the articles and its categories.
If there a second call for `['article.2', 'article.4]` then will be preloaded only the articles.

While SQL will do "full query" for each case. Sounds less efficient.

Fedik committed:
Access::preloadItems

Fedik commented:
> Is it faster if you add these indexes?
> ```ALTER TABLE `#__assets` ADD INDEX `idx_lft` (`lft`), ADD INDEX `idx_rgt` (`rgt`);```

Nope, unfortunately. Before applying the index it took ~250ms, and after ~250ms.

It probably skipping index because of range check:
![Screenshot 2024-09-14_14-09-28](https://github.com/user-attachments/assets/9cfd1799-8a73-487a-8a71-85236b2889be)

Fedik committed:
shorten

brianteeman commented on administrator/components/com_categories/src/View/Categories/HtmlView.php:
```diff
@@ -119,11 +120,22 @@ public function display($tpl = null)
+        foreach ($this->items as $item) {
             $this->ordering[$item->parent_id][] = $item->id;
+
+            $assetsList['c' . $item->id] = 'com_content.category.' . $item->id;
+
+            if (!empty($item->parent_id)) {
+                $assetsList['c' . $item->parent_id] = 'com_content.category.' . $item->parent_id;
```
what about when com_categories is used with any other component that supports categories and not just com_content

Fedik replied:
Good point. Fixed it. Thanks!

Fedik committed:
category component

brianteeman commented on administrator/components/com_categories/src/View/Categories/HtmlView.php:
```diff
@@ -119,11 +120,23 @@ public function display($tpl = null)
+            }
         }
 
+        Access::preloadItems('com_content', array_values($assetsList));
```
as before - what about other components that use com_categories

Fedik replied:
The same :smile: 
Thanks!

Fedik committed:
category component

Quy commented on libraries/src/Access/Access.php:
```diff
@@ -209,6 +209,57 @@ public static function preload($assetTypes = 'components', $reload = false)
+     * @param  array   $assetsList     Assets list. Either list of asset names or asset ids.
+     * @param  string  $key            The key to use in the filter:
+     *                                 id - When $assetsList is list of asset ids;
+     *                                 name - When $assetsList is list of asset names;
+     *
+     * @return void
+     */
```
```suggestion
     * @return void
     * 
     * @since   __DEPLOY_VERSION__
     */
```

Fedik committed:
Update libraries/src/Access/Access.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

Fedik committed:
Merge branch '5.3-dev' into access-smart-preload

Fedik committed:
cs

laoneo committed:
Merge branch '5.3-dev' into access-smart-preload

HLeithner commented:
This pull request has been automatically rebased to 6.0-dev.

Fedik committed:
Merge branch '6.0-dev' into access-smart-preload

Fedik commented:
@SniperSister please have a look also this one. In combination with https://github.com/joomla/joomla-cms/pull/45542 it should be a bigger improvement in performance for the Content component for large sites.

Fedik committed:
Merge branch '6.0-dev' into access-smart-preload

SniperSister commented on libraries/src/Access/Access.php:
```diff
@@ -209,6 +209,59 @@ public static function preload($assetTypes = 'components', $reload = false)
+        }
+
+        if ($pids) {
+            // Make sure parents also loaded
```
Depending on the nesting level, we'll end up with quite a number of calls here. What was the query that you've been using to retreive the ancestors based on the nested sets structure?

Fedik replied:
It was LEFT JOIN with lft, rgt stuff.
You can find discussion in hidden "review" https://github.com/joomla/joomla-cms/pull/44078#discussion_r1759710276
And my conclusion https://github.com/joomla/joomla-cms/pull/44078#discussion_r1759991908

In short:
The parents in the tree is cached, so even when code will call this recursion it still will load the parents only once.
Within reasonable limits the multiple queries is faster than one huge (which always will load full tree even if it was loaded previously).

Fedik committed:
Merge branch '6.0-dev' into access-smart-preload

HLeithner commented:
This pull request has been automatically rebased to 6.1-dev.

Fedik committed:
Merge branch '6.1-dev' into access-smart-preload

 Conflicts:
	administrator/components/com_categories/src/View/Categories/HtmlView.php

Fedik committed:
Sync. Merge branch '6.2-dev' into access-smart-preload

 Conflicts:
	components/com_content/src/Model/ArticleModel.php
	components/com_content/src/View/Category/HtmlView.php
